### PR TITLE
Qr environment fix

### PR DESCRIPTION
**Pull request by Oktober13**
Contains:
* QR Code Gazebo materials & models
* Code to automatically update the qr code to a preexisting qr code in the QR Codes folder
* Code to generate new qr codes in the QR Codes folder
* Updates to existing QR code generation code for clarity

### CODE REVIEW

File: iarc_sim_3d/media/materials/textures/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+*.png
```
Comment (yycho0108) [inline iarc_sim_3d/media/materials/textures/.gitignore]: How did 0140-*.png get committed despite this .gitignore?

File: iarc_fuses/scripts/QRGenerator.py [hunk 2]
```diff
@@ -1,25 +1,111 @@
...
+import argparse
 import os
+import random
+import re
 import rospkg
 from random import randint
+from shutil import copyfile
+import qrcode
+
+class qrGenerator(object):
```
Comment (yycho0108) [inline iarc_fuses/scripts/QRGenerator.py]: For me, QRGenerator() feels like a more conventional (ROS-compliant) class name, but it's really not that important.

File: iarc_fuses/scripts/QRGenerator.py [hunk 3]
```diff
@@ -1,25 +1,111 @@
...
+    def __init__(self, args):
+        rospack = rospkg.RosPack() 
+        pkgRoot = rospack.get_path('iarc_fuses') # Gets the package
+        mydir = os.path.join(pkgRoot,"QRCodes")
+
+        if args.values()[0]:
```
Comment (yycho0108) [inline iarc_fuses/scripts/QRGenerator.py]: Unclear why you're checking for args.values()[0] -- what does this line mean?

File: iarc_fuses/scripts/QRGenerator.py [hunk 4]
```diff
@@ -1,25 +1,111 @@
...
+                code = '{0:04}'.format(randint(0, 10**3))
+                image = qrGenerator.generateQR(pkgRoot, code)
+                qrGenerator.cropQR(pkgRoot, image, code)
+
+            qrGenerator.setQRMaterial(rospack, code)
+        else:
+            filelist = [ f for f in os.listdir(mydir) if f.endswith(".png") ]
+            rand = random.randint(0,len(filelist)-1)
+            print rand
+            code = filelist[rand].replace('-1.png', '').replace('-2.png', '').replace('-3.png', '').replace('-4.png', '').replace('.png', '')
```
Comment (yycho0108) [inline iarc_fuses/scripts/QRGenerator.py]: Interesting way to deal with filename parsing... consider regex?

File: iarc_fuses/scripts/QRGenerator.py [hunk 5]
```diff
@@ -1,25 +1,111 @@
...
+        for i in range(2):
+            for j in range(2):
+                box = (imgWidth / 2 * i, int(imgHeight / 2 * j),imgWidth / 2 * (1+i), imgHeight / 2 * (1+j)) # selects a box to crop
+                quart = img.crop(box) # crops the image into one quarter
+                section = str(1 + i + j*2) # the section number of the image, 1 = top left, 2 = top right, 3 = bottom left, 4 = bottom right
+                quart.save(os.path.join(pkgRoot,"QRCodes", code + "-" + section + '.png'),"PNG") # saves the quarter image
+        return
+
+    @staticmethod
+    def replaceFilename(filepath, code, num):
```
Comment (yycho0108) [inline iarc_fuses/scripts/QRGenerator.py]: Why is this function not used above?

File: iarc_sim_3d/media/materials/scripts/iarc.material [hunk 6]
```diff
@@ -93,3 +93,91 @@ material IARC/TopGreen{
...
+    technique
+    {
+        pass
+        {
+            ambient 0.5 0.5 0.5 1.0
+            diffuse 0.5 0.5 0.5 1.0
+            specular 0.2 0.2 0.2 1.0 12.5
+            texture_unit
+            {
+                texture 0140-1.png
```
Comment (yycho0108) [inline iarc_sim_3d/media/materials/scripts/iarc.material]: Did you resolve the image repeating issue? (i.e. only one texture unit should appear on the surface at full scale, not a repeated tile)

File: iarc_sim_3d/worlds/iarc.world [hunk 7]
```diff
@@ -35,6 +35,18 @@
         <include>
             <uri>model://iarc_sim_3d/models/floor</uri>
         </include>
+        <include>
+            <uri>model://iarc_sim_3d/models/qr1</uri>
```
Comment (yycho0108) [inline iarc_sim_3d/worlds/iarc.world]: Potentially place these QR codes on top of the bins or something

File: iarc_sim_3d/media/materials/textures/.gitignore (hunk 1 above)
Comment (Oktober13) [reply iarc_sim_3d/media/materials/textures/.gitignore]: It's intentional, so code will run out of the box when people pull (without having to run QR code generation immediately.)

File: iarc_fuses/scripts/QRGenerator.py (hunk 3 above)
Comment (Oktober13) [reply iarc_fuses/scripts/QRGenerator.py]: This is checking whether the argument -n or --new was added, which causes all QR codes currently in the folder to be deleted and regenerated.
Argparse values is a list variable that contains all inputted arguments, but I only care about -n.

Comment (Oktober13) [reply iarc_fuses/scripts/QRGenerator.py]: Adding comment for clarity

File: iarc_sim_3d/media/materials/scripts/iarc.material (hunk 6 above)
Comment (Oktober13) [reply iarc_sim_3d/media/materials/scripts/iarc.material]: The image tiling issue is resolved. The QR code is 0.1 m x 0.1 in size.

File: iarc_fuses/scripts/QRGenerator.py (hunk 5 above)
Comment (Oktober13) [reply iarc_fuses/scripts/QRGenerator.py]: It's a helper function used in setQRMaterial, which is the broader function that handles replacing the filepath for all 4 QR quarter images.

**Comment by Oktober13** [comment]
Sim.launch is refusing to launch now; may require more effort.

**Comment by Oktober13** [comment]
QR quarters have been added to sim. Unfortunately, couldn't link with the dumpsters directly because they use a .dae, but they're hovering over the three dumpsters.

File: iarc_sim_3d/models/qr1/model.sdf [hunk 8]
```diff
@@ -0,0 +1,54 @@
+<?xml version="1.0"?>
+<sdf version="1.4">
+    <model name="QR1">
+        <static>true</static>
+        <link name="link">
+            <collision name="collision">
+                <geometry>
+                    <plane>
+                        <normal>0 0 2</normal>
+                        <size>0.1 0.1</size>
```
Comment (yycho0108) [inline iarc_sim_3d/models/qr1/model.sdf]: Is this size roughly consistent with what we expect?

**Pull request merged**